### ISSUE TITLE
terraform import support for main resources (contabo_instance)

### ISSUE THREAD

**Issue by brahima** [contabo/terraform-provider-contabo#3]
Hello,

Just bought a VPS on contabo and i'd like to use terraform.
The provider does not seem to support the **terraform import** command for main resources (mainly _contabo_instance_)

Am i missing something or the only way to use terraform is to create an instance through it ?

Thanks for you help
Regards,

Brahim

**Comment by tahergalal** [comment]
Hi,

For the time being there is actually still no import this is planned to be released but to handle instance niw with terraform youcwould need to create them through terraform

**Comment by brahima** [comment]
Hi @tahergalal,
Ok thanks for your quick reply.

There's no _hack_ to generate a tf state file for my VPS or May be i can contact the support to see what can be done ?

Thanks

**Comment by tahergalal** [comment]
I am not sure if there is a hack/workaround will check and come back to you as soon as possible

**Comment by tahergalal** [comment]
so what you can try to do as a workaround is to create the terrform.tfstate file yourself and fill in the values of the instance your self by doing a get using curl. You can then try to use it normally can't guarantee that this works 100% but should be a starting point

```json
{
  "version": 4,
  "terraform_version": "1.2.5",
  "serial": 16,
  "lineage": "9e271590-e79b-b8e6-cdce-50a882df29af",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "contabo_instance",
      "name": "name you want to use",
      "provider": "provider[\"registry.terraform.io/contabo/contabo\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "add_ons": [],
            "cancel_date": "",
            "cpu_cores": 4,
            "created_date": "Saturday, 03-Sep-22 00:11:46 UTC",
            "disk_mb": 204800,
            "display_name": "",
            "error_message": "",
            "id": "instance id",
            "image_id": "image id",
            "ip_config": [
              {
                "v4": [
                  {
                    "gateway": "your ip config",
                    "ip": "your ip config",
                    "netmask_cidr": your ip config
                  }
                ],
                "v6": [
                  {
                    "gateway": "your ip config",
                    "ip": "your ip config",
                    "netmask_cidr": your ip config
                  }
                ]
              }
            ],
            "last_updated": null,
            "license": null,
            "mac_address": "00:50:56:47:c5:0f",
            "name": "your instance name",
            "os_type": "type from api",
            "period": null,
            "product_id": "prpoduct from api",
            "product_type": "product type",
            "ram_mb": 8192,
            "region": "EU",
            "root_password": null,
            "ssh_keys": [],
            "status": "running",
            "user_data": null,
            "v_host_id": your vhost id
          },
          "sensitive_attributes": [],
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}


```


obviously you will need to edit the values above to work with your instance maybe you can tell me if this works.

**Comment by brahima** [comment]
Hello @tahergalal ,

It worked. Here is what i did (maybe that will help someone) : 

**Note** : I'm using Terraform cloud (terraform.io)

- i declared a data contabo_instance

```
data "contabo_instance" "inst" {
  id = "my_instance_id"
}

output "instance_id" {
  description = "ID of the contabo instance"
  value       = data.contabo_instance.inst.id
}
```

- i played the terraform plan / apply command to fill my state with the data resource and all the attributes associated with my instance (to avoid copy/pasting all the info)

- i saved and backed up my state with : terraform state pull > my_state_file.tfstate then cp to backup it

- i then manually edited my state : 
  - i incremented the serial property of the state file 
  - copy/pasted the contabo_instance data resource previously added and 
    - replaced mode = data with mode managed
    - changed the name (but it can be the same)
    - checked all the other fields (attributes) ... but all seemed good to me as far as i remember
   
- i made sure that my state file was a valid json file and played : terraform state push my_state_file.tfstate

- I added the resource to my tf file so that _terraform plan_ indicates no changes

```
resource "contabo_instance" "instance" {
  product_id = "V13"
  region     = "EU"
}
```

**Comment by brahima** [comment]
@tahergalal 
I now have a problem trying to use cloud-init (via the user_data field of the contabo_instance resource).

I succeeded to re-install my instance using cloud-init via the contabo UI and via the API.

But via terraform, i keep having the following error :

**API error, status code: 400, details: json: cannot unmarshal array into Go struct field ApiError.message of type string{"statusCode":400,"message":["imageId must be a UUID"]}**

If i remove the [user_data](https://registry.terraform.io/providers/contabo/contabo/latest/docs/resources/contabo_instance#user_data) field, my terraform apply is ok and i can re-install my machine (i did some test switching the image_id).

So my question is : **Do you have a working exemple (even simple) with user_data field (cloud-init) ?**

Here is what i have :

```
data "contabo_image" "debian_11" {
  id = "66abf39a-ba8b-425e-a385-8eb347ceac10"
}

resource "contabo_instance" "preview_1" {
  image_id = data.contabo_image.debian_11.id
  period     = 1
  product_id = "V13"
  region     = "EU"
  user_data = file("${path.module}/cloud-init/cloud.cfg.tftpl")
}
```

My config file cloud.cfg.tftpl contains :  (i tested a lot of things ... without any success)
```
#cloud-config
runcmd:
  - ls /
```

Thanks for your help

**Comment by brahima** [comment]
Hello,

@tahergalal any thoughts about my last issue. 
I did some other tests without any success. I tried to format the user_data content removing the line return ... like i do via an HTTP client ... to no avail.

Any help would be greatly appreciated ... 

Best regards,
Brahim

**Comment by tahergalal** [comment]
Will start testing will come back to you as soon as I have a sol. Will also check if this is not just a bug in the terraform provider and  will require a fix

**Comment by brahima** [comment]
Ok thanks for you help. Keep me updated if you want me to test things on my side.

**Comment by tahergalal** [comment]
this should be fixed now with the new version of terraform 0.1.13 please use this version and if there are other problems create a new issue.
